tests/cpp_validation_tests/test_validation.cpp

```diff
@@ -2,27 +2,70 @@
 //
 // SPDX-License-Identifier: MPL-2.0
 
+#define PGM_ENABLE_EXPERIMENTAL
+
+#include "power_grid_model_cpp.hpp"
+
 #include <power_grid_model/auxiliary/dataset.hpp>
 #include <power_grid_model/auxiliary/meta_data_gen.hpp>
 #include <power_grid_model/auxiliary/serialization/deserializer.hpp>
 #include <power_grid_model/container.hpp>
 #include <power_grid_model/main_model.hpp>
 
+#include <Eigen/Dense>
 #include <doctest/doctest.h>
 #include <nlohmann/json.hpp>
 
+#include <complex>
 #include <concepts>
 #include <cstdlib>
 #include <cstring>
 #include <filesystem>
 #include <fstream>
 #include <iostream>
+#include <limits>
 #include <optional>
 #include <regex>
+#include <stdexcept>
+#include <type_traits>
 
 namespace power_grid_model::meta_data {
 
 namespace {
+// taken from common.hpp
+using IdxVector = std::vector<Idx>;
+
+// taken from three_phase_tensor.hpp
+template <class T> using Eigen3Vector = Eigen::Array<T, 3, 1>;
+
+// is nan. Taken from three_phase_tensor.hpp
+template <class Derived> inline bool is_nan(Eigen::ArrayBase<Derived> const& x) { return x.isNaN().all(); }
+inline bool is_nan(std::floating_point auto x) { return std::isnan(x); }
+template <std::floating_point T> inline bool is_nan(std::complex<T> const& x) {
+    return is_nan(x.real()) || is_nan(x.imag());
+}
+inline bool is_nan(int32_t x) { return x == std::numeric_limits<int32_t>::min(); }
+inline bool is_nan(int8_t x) { return x == std::numeric_limits<int8_t>::min(); }
+template <class Enum>
+    requires std::same_as<std::underlying_type_t<Enum>, int8_t>
+inline bool is_nan(Enum x) {
+    return static_cast<int8_t>(x) == std::numeric_limits<int8_t>::min();
+}
+
+template <class Functor, class... Args> decltype(auto) pgm_type_func_selector(Idx type, Functor&& f, Args&&... args) {
+    switch (type) {
+    case Idx{0}: // PGM_int32
```

Comment:
why use `Idx{0}`? What's the problem to use `PGM_int32` directly?

Reply:
Because the attribute type as returned by `power_grid_model_cpp::MetaData::attribute_ctype` is a raw C `enum`.

Reply:
That should not hurt right? A raw C enum is just an integer literal, you can use it directly.

Reply:
This has been addressed now.